Implement event firing from the web process for webRequest events and add tests

b-weinstein opened the pull request:
#### 0c0655a175613f58168c7f30652e989cc0fe2736
<pre>
Implement event firing from the web process for webRequest events and add tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=267963">https://bugs.webkit.org/show_bug.cgi?id=267963</a>
<a href="https://rdar.apple.com/114823223">rdar://114823223</a>

Reviewed by Timothy Hatcher.

The tests exposed a few issues that were also fixed here:
1) The network process wasn&apos;t calling into WebPageProxy when running the tests because the
page didn&apos;t have a resource load delegate. To fix this, also call into the WebPageProxy when
the page has an extension controller.
2) There was a copy/paste-o in WebExtensionContext::resourceLoadDidPerformHTTPRedirection where
the wrong events were being sent.
3) When just a filter was being passed to webRequest.event.addListener, it was being treated as the extraInfoSpec because
we weren&apos;t being specific enough about the type.

This PR also changed _WKWebExtensionWebRequestFilter to deal with WebKit::ResourceLoadInfo instead of _WKResourceLoadInfo.

* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters): Add a new pageHasExtensionController parameter.
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldSendResourceLoadMessages const): Check both pageHasResourceLoadClient and pageHasExtensionController.
(WebKit::NetworkResourceLoader::startNetworkLoad): Adopt shouldSendResourceLoadMessages.
(WebKit::NetworkResourceLoader::didReceiveResponse): Ditto.
(WebKit::NetworkResourceLoader::didFinishLoading): Ditto.
(WebKit::NetworkResourceLoader::didFailLoading): Ditto.
(WebKit::NetworkResourceLoader::didReceiveChallenge): Ditto.
(WebKit::NetworkResourceLoader::continueWillSendRequest): Ditto.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h:
(WebKit::toWebExtensionFrameIdentifier): Add a new flavor of this method.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::resourceLoadDidPerformHTTPRedirection): Make sure we are calling the correct events here.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::convertRequestBodyToWebExtensionFormat): Convert the request body to the format extensions expect.
(WebKit::webRequestDetailsForResourceLoad): Generate a dictionary of information about the resource load conforming to the webRequest spec.
(WebKit::convertHeaderFieldsToWebExtensionFormat): Convert the header fields to an array of dictionaries.
(WebKit::headersReceivedDetails): Convert the NSHTTPURLResponse information into a dictionary conforming to the webRequest spec.
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest): Call the correct events.
(WebKit::WebExtensionContextProxy::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestEventCocoa.mm:
(WebKit::WebExtensionAPIWebRequestEvent::invokeListenersWithArgument): Check the filter and invoke the correct listeners.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebRequestEvent.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebRequestFilter.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebRequestFilter.mm:
(_WKWebExtensionWebRequestResourceTypeFromResourceLoadInfo): Updated to deal with ResourceLoadInfo::Type.
(_WKWebExtensionWebRequestResourceTypeFromWKResourceLoadInfo): Deleted.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebRequestEvent.idl: Make sure extraInfoSpec is an array of strings.
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::addParametersShared): Set pageHasExtensionController.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
(TestWebKitAPI::TEST): Add tests for various parts of webRequest functionality.

Canonical link: <a href="https://commits.webkit.org/273393@main">https://commits.webkit.org/273393@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c652e1154d03937d21919e2ce96d58e47423ca5d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14289 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38100 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/31891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16660 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11334 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35912 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/12063 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/31489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10591 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39346 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32124 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31937 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10791 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8696 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34647 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8075 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11315 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/11604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
```diff
@@ -124,29 +127,177 @@
+        }
+        [array addObject:item.value];
+    }
+    return dictionary;
```
Add newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
```diff
@@ -124,29 +127,177 @@
+        },
+        @"isProxy": @(challenge.protectionSpace.isProxy),
+    }];
+    if (NSString *realm = challenge.protectionSpace.realm)
```
Add newline above.

xeenon commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -345,6 +345,11 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+        if (auto* webPage = WebPage::fromCorePage(*page))
+            parameters.pageHasExtensionController = !!webPage->webExtensionControllerProxy();
```
`!!` isn't needed for C++ `bool`.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
```diff
@@ -31,7 +31,7 @@
 namespace TestWebKitAPI {
 
-static auto *webRequestManifest = @{ @"manifest_version": @3, @"permissions": @[ @"webRequest" ], @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @NO } };
+static auto *webRequestManifest = @{ @"manifest_version": @3, @"permissions": @[ @"webRequest" ], @"background": @{ @"scripts": @[ @"background.js" ], @"type": @"module", @"persistent": @YES } };
```
You should make this whole file `#if PLATFORM(MAC)` and eventually add a test that shows it isn't available on iOS. The `persistent` will trigger an error on iOS.

b-weinstein replied:
That sounds good. I'll tackle this in a follow-up - there are a number of places I want to do this.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
```diff
@@ -53,6 +53,254 @@
+        @"function beforeSendHeadersHandler() { beforeSendHeadersFired = true }",
+        @"function sendHeadersHandler() { sendHeadersFired = true }",
+        @"function headersReceivedHandler() { headersReceivedFired = true }",
+        @"function responseStartedHandler() { responseStartedFired = true }",
```
These would be clearer inline with `addListener` below.

```
browser.webRequest.onBeforeRequest.addListener(() => { beforeRequestFired = true })
```

b-weinstein replied:
Way clearer.

b-weinstein committed:
Implement event firing from the web process for webRequest events and add tests
https://bugs.webkit.org/show_bug.cgi?id=267963
rdar://114823223

Reviewed by Timothy Hatcher.

The tests exposed a few issues that were also fixed here:
1) The network process wasn't calling into WebPageProxy when running the tests because the
page didn't have a resource load delegate. To fix this, also call into the WebPageProxy when
the page has an extension controller.
2) There was a copy/paste-o in WebExtensionContext::resourceLoadDidPerformHTTPRedirection where
the wrong events were being sent.
3) When just a filter was being passed to webRequest.event.addListener, it was being treated as the extraInfoSpec because
we weren't being specific enough about the type.

This PR also changed _WKWebExtensionWebRequestFilter to deal with WebKit::ResourceLoadInfo instead of _WKResourceLoadInfo.

* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters): Add a new pageHasExtensionController parameter.
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::shouldSendResourceLoadMessages const): Check both pageHasResourceLoadClient and pageHasExtensionController.
(WebKit::NetworkResourceLoader::startNetworkLoad): Adopt shouldSendResourceLoadMessages.
(WebKit::NetworkResourceLoader::didReceiveResponse): Ditto.
(WebKit::NetworkResourceLoader::didFinishLoading): Ditto.
(WebKit::NetworkResourceLoader::didFailLoading): Ditto.
(WebKit::NetworkResourceLoader::didReceiveChallenge): Ditto.
(WebKit::NetworkResourceLoader::continueWillSendRequest): Ditto.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/Shared/Extensions/WebExtensionFrameIdentifier.h:
(WebKit::toWebExtensionFrameIdentifier): Add a new flavor of this method.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::resourceLoadDidPerformHTTPRedirection): Make sure we are calling the correct events here.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::convertRequestBodyToWebExtensionFormat): Convert the request body to the format extensions expect.
(WebKit::webRequestDetailsForResourceLoad): Generate a dictionary of information about the resource load conforming to the webRequest spec.
(WebKit::convertHeaderFieldsToWebExtensionFormat): Convert the header fields to an array of dictionaries.
(WebKit::headersReceivedDetails): Convert the NSHTTPURLResponse information into a dictionary conforming to the webRequest spec.
(WebKit::WebExtensionContextProxy::resourceLoadDidSendRequest): Call the correct events.
(WebKit::WebExtensionContextProxy::resourceLoadDidPerformHTTPRedirection): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveChallenge): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveResponse): Ditto.
(WebKit::WebExtensionContextProxy::resourceLoadDidCompleteWithError): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestEventCocoa.mm:
(WebKit::WebExtensionAPIWebRequestEvent::invokeListenersWithArgument): Check the filter and invoke the correct listeners.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebRequestEvent.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebRequestFilter.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebRequestFilter.mm:
(_WKWebExtensionWebRequestResourceTypeFromResourceLoadInfo): Updated to deal with ResourceLoadInfo::Type.
(_WKWebExtensionWebRequestResourceTypeFromWKResourceLoadInfo): Deleted.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebRequestEvent.idl: Make sure extraInfoSpec is an array of strings.
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::addParametersShared): Set pageHasExtensionController.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIWebRequest.mm:
(TestWebKitAPI::TEST): Add tests for various parts of webRequest functionality.

Canonical link: https://commits.webkit.org/273393@main
